src/connectors/impls/gpubsub/consumer.rs

```diff
@@ -0,0 +1,399 @@
+// Copyright 2022, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::connectors::google::AuthInterceptor;
+use crate::connectors::prelude::*;
+use crate::connectors::utils::url::HttpsDefaults;
+use async_std::channel::{Receiver, Sender};
+use async_std::stream::StreamExt;
+use async_std::task;
+use async_std::task::JoinHandle;
+use beef::generic::Cow;
+use dashmap::DashMap;
+use googapis::google::pubsub::v1::subscriber_client::SubscriberClient;
+use googapis::google::pubsub::v1::{PubsubMessage, ReceivedMessage, StreamingPullRequest};
+use gouth::Token;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::Duration;
+use tonic::codegen::InterceptedService;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Status;
+use tremor_pipeline::ConfigImpl;
+
+#[derive(Deserialize, Clone)]
+struct Config {
+    #[serde(default = "default_connect_timeout")]
+    pub connect_timeout: u64,
+    #[serde(default = "default_ack_deadline")]
+    pub ack_deadline: u64,
+    pub subscription_id: String,
+    #[serde(default = "default_endpoint")]
+    pub endpoint: String,
+    #[cfg(test)]
+    #[serde(default = "default_skip_authentication")]
+    pub skip_authentication: bool,
+}
+impl ConfigImpl for Config {}
+
+fn default_endpoint() -> String {
+    "https://pubsub.googleapis.com".into()
+}
+
+#[cfg(test)]
+fn default_skip_authentication() -> bool {
+    false
+}
+
+fn default_connect_timeout() -> u64 {
+    1_000_000_000u64 // 1 second
+}
+
+fn default_ack_deadline() -> u64 {
+    10_000_000_000u64 // 10 seconds
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "gpubsub-consumer".into()
```

Comment:
Only a tiny thing, I prefer the `-` writing but we don't allow `-` in idents so anyone would have to escape this when using it, therefore we use `_` in the connector names (as ugly as it looks :sob:)
```suggestion
        "gpubsub_consumer".into()
```